aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,186 @@
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+
+namespace at {
+namespace autocast {
+
+namespace {
+// Establish the same cast-caching policy as Apex (cache FP16 casts of FP32 leaves).
+// The key type is a proxy for a Tensor uuid.
+// The value type is Tensor to make sure cached Tensors stay alive.
+thread_local std::unordered_map<TensorImpl*, Tensor> cached_leaf_casts;
```

Comment:
Maybe you want to use a weak pointer here?

Reply:
I lean towards no because t's possible that for some cached casts, the cast cache is the _only_ thing keeping the casted tensor alive.

Reply:
The key. TensorImpl* is "weak" but it's technically unsafe if TensorImpl gets deallocated and then reallocated somewhere else. WeakTensor won't have this problem.

Reply:
WeakTensor doesn't appear to exist.  Implemented the fix we talked about in Slack and added comments:  https://github.com/pytorch/pytorch/pull/29552/files#diff-e955d2ce1fd0d1e4faeb843a08a3f324R9-R26